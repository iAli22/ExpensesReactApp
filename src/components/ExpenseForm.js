import React, { Component } from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";

class ExpenseForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calandarFocused: false,
      errors: ""
    };
  }

  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
    //this.setState({ [e.target.name]: e.target.value });
  };

  onNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };

  onAmountChange = e => {
    const amount = e.target.value;

    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      //regex
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = createdAt => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calandarFocused: focused }));
  };

  onSubmit = e => {
    e.preventDefault();
    const { description, amount, createdAt, note } = this.state;
    // validation
    if (!description || !amount) {
      this.setState(() => ({
        errors: "Please set Provide Descriptipn and Amount"
      }));
    } else {
      //Clear Error
      this.setState(() => ({ errors: "" }));
      this.props.onSubmit({
        description,
        amount: parseFloat(amount, 10) * 100,
        createdAt: createdAt.valueOf(),
        note
      });
    }
  };
  render() {
    const { errors } = this.state;
    return (
      <div>
        {errors !== "" ? <p style={{ color: "red" }}>{errors}</p> : null}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            // name="description"
            placeholder="Type Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="number"
            name="amount"
            placeholder="Type Amount"
            onChange={this.onAmountChange}
            value={this.state.amount}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calandarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={day => false}
          />
          <textarea
            name="note"
            placeholder="Add Note For Your Expense"
            value={this.state.note}
            onChange={this.onNoteChange}
          />
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}

export default ExpenseForm;
