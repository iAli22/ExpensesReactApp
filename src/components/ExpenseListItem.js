import React from "react";
import { Link } from "react-router-dom";

export const ExpenseListItem = ({ id, description, amount, createdAt }) => (
  <div>
    <h3>{description}</h3>
    <p>
      {amount} - {createdAt}
    </p>

    <button>
      <Link to={`/edit/${id}`}>Edit</Link>
    </button>
  </div>
);

export default ExpenseListItem;
