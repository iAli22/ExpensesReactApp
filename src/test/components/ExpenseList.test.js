import React from "react";
import { shallow } from "enzyme";
import { ExpenseList } from "../../components/ExpenseList";
import expenses from "../fixtures/expenses";

test("Should Render ExpenseList with expenses", () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />);
  expect(wrapper).toMatchSnapshot();
});

test("Should Render ExpenseList With Empty Message", () => {
  const wrapper = shallow(<ExpenseList expenses={[]} />);
  expect(wrapper).toMatchSnapshot();
});
