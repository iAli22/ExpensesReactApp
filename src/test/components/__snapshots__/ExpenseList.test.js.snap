// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should Render ExpenseList With Empty Message 1`] = `
<div>
  <p>
    There Is No Expenses To Show
  </p>
  <h1>
    Expense List
  </h1>
</div>
`;

exports[`Should Render ExpenseList with expenses 1`] = `
<div>
  <ExpenseListItem
    amount={195}
    createdAt={0}
    description="Gum"
    id="1"
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={109500}
    createdAt={-345600000}
    description="Rent"
    id="2"
    key="2"
    note=""
  />
  <ExpenseListItem
    amount={4500}
    createdAt={345600000}
    description="Credit Card"
    id="3"
    key="3"
    note=""
  />
  <h1>
    Expense List
  </h1>
</div>
`;
